### ISSUE TITLE
Not possible to set disabled state to a tree view item

### ISSUE THREAD

**Issue by victorg1991** [liferay/clay#4676]
Right now, (at least from what I have seen, I may be mistaken) it is not possible to set the disabled state for a tree view item. It would be nice to have a property in the `<ClayTreeView.ItemStack>` or  `<ClayTreeView.Item>` to set this state.

**Comment by matuzalemsteles** [comment]
Hey @victorg1991 what use for `disabled` in this case would be just to disable Node and not its children, correct? or is it a case of disabling Node and its children? Currently in the implementation adding `disabled` also causes the component not to be expanded if it has children. @pat270 maybe now we need to add style to `disabled` in TreeView Node 😄

**Comment by victorg1991** [comment]
Yeah, correct, ideally children should be enabled even if the parent is not. An use case may be selecting a folder to where a document should be moved, you may want to disable the current folder of the document, but not its children :)

**Comment by victorg1991** [comment]
> Currently in the implementation adding disabled also causes the component not to be expanded if it has children. @pat270 maybe now we need to add style to disabled in TreeView Node 😄

I tried that without luck, maybe I wasn't used the latest clay css version, in which component did you add the disabled class?

**Comment by victorg1991** [comment]
I guess in those case we also want to ignore the click event in the item?

**Comment by matuzalemsteles** [comment]
Hey @victorg1991 this behavior only worked for Drag and Drop, you can pass the property it will "work" but it will not disable anything, we will have to add the CSS for this case, maybe we have to rethink because I think we didn't expect this case to still keep expanding even though the Node is disabled.

@drakonux maybe we need to add this use case in the documentation as well.

> I guess in those case we also want to ignore the click event in the item?

Probably when adding the disabled properties in the CSS will do it for us. But today you can do that.

```jsx
<TreeView.ItemStack
  onClick={(event) => item.disabled && event.preventDefault()}
>

<TreeView.Item
  onClick={(event) => item.disabled && event.preventDefault()}
>
```

In this case, it will depend on how your `selectionMode` is, for example in the single selection, it will ignore the selection and expand when you click on Node but clicking on the expand icon will continue to work.

When your selection is multiple or multiple-recursive, as it is via Checkbox, in this case above it will not be necessary because it will still select but you can also ignore it in two ways.

```jsx
<TreeView.ItemStack>
  <Checkbox
    disabled={item.disabled}
  />

  <Checkbox
    onChange={(event) => item.disabled && event.preventDefault()}
  />

  <Checkbox
    readOnly
  />
</TreeView.ItemStack>
```

Both cases have different behavior, the first will ignore the selection and visually show the disabled checkbox, the second will just ignore the selection but the user can click on the checkbox, the third is similar to the second.

Looking at these @victorg1991 cases, would that be enough? I'm imagining that adding a `disabled` for example in `TreeView.Item` or `TreeView.ItemStack` will be inconsistent with the idea of disabling Node and expander, it will be a bit tricky to get this to work OOTB for both cases without providing something an auxiliary prop and I wanted to avoid that for this scenario. What do you think?

**Comment by pat270** [comment]
We can add `disabled` to `treeview-item` for disabling everything inside. If you only want to disable a specific node, we can add it to the `treeview-link`. This isn't implemented yet, but what I was thinking we could do should this need arise.

**Comment by victorg1991** [comment]
@matuzalemsteles I think that functionality wise this will be enough, but how we can set the disabled look and feel? Passing the "disabled" class would be enough?

> We can add disabled to treeview-item for disabling everything inside. If you only want to disable a specific node, we can add it to the treeview-link. This isn't implemented yet, but what I was thinking we could do should this need arise.
That sounds great 😍

**Comment by drakonux** [comment]
Hello everyone, 

We have disabled states for nodes in the documentation, but probably we didn't cover all the cases that the tree view nodes have to support. So could you @matuzalemsteles create us a ticket for the cases we should add, please?

Kind regards,
David

**Comment by matuzalemsteles** [comment]
> I think that functionality wise this will be enough, but how we can set the disabled look and feel? Passing the "disabled" class would be enough?

Probably so, I'm still worried that adding `disabled` won't affect the expander, because adding `disabled` to the `treeview-link` will also visually disable the Expander, because they're all on the same element, a more simplified example of markup:

```html
<div className="treeview-link">
  <Layout.ContentRow>
    <Layout.ContentCol>
      Expander here
    </Layout.ContentCol>

    <Layout.ContentCol>
      <div className="component-icon">Icon</div>
    </Layout.ContentCol>

    <Layout.ContentCol expand>
      <div className="component-text">Name</div>
    </Layout.ContentCol>
  </Layout.ContentRow>
</div>
```

@pat270 for that case do you see any way to bypass the expander?

> So could you @matuzalemsteles create us a ticket for the cases we should add, please?

Of course, once we get an idea of how this is going to work here, I'll create a ticket.

**Comment by pat270** [comment]
@matuzalemsteles we will need to add `disabled` on the `treeview-link` and each `button` and `a` element inside that we want disabled. The expander won't have `disabled`. The catch is we will need `disabled` on every nested `treeview-link`, not sure if this is even relevant.

**Comment by julien** [comment]
> @matuzalemsteles we will need to add disabled on the treeview-link and each button and a element inside that we want disabled.

Thanks @pat270, seems a bit complex inmho.

**Comment by matuzalemsteles** [comment]
Hey @victorg1991 the new release of Clay should include this fix now, we no longer need to ignore or avoid calling some method, so at the implementation level, you just need to do it.
```jsx
{(item) => (
  <TreeView.Item>
    <TreeView.ItemStack disabled={item.disabled} expanderDisabled={false}>
     ...
    </TreeView.ItemStack>
    <TreeView.Group items={item.children}>
      {(item) => (
        <TreeView.Item disabled={item.disabled}>
          ..
        </TreeView.Item>
      )}
    </TreeView.Group>
  </TreeView.Item>
)}
```

`expanderDisabled` will make the Expander stay active and not respect the `disabled` property. There's more information in the documentation too, feel free to provide feedback.

**Comment by victorg1991** [comment]
Nice 👏  Thanks @matuzalemsteles :)
